### PR TITLE
Implement the SafeLogger API

### PR DESCRIPTION
**Pull request by carterkozak**
This API is meant to enforce safe-logging concepts at compile-time
in an obvious way, without reliance on retrofitting error-prone
checks to inform usage.

A few key concepts:
* `Arg` types are always required
* message strings are annotated `CompileTimeConstant` out of the box.
* The `SafeLogger` type cannot be subtyped. Arbitrary implementations
  cannot be created and passed becuase they often result in
  unintentional side-effects.
* Markers are not supported.
* Events logged using this API may be differentiated from
  other events by layout implementations. The slf4j implementation
  handles this using markers.
* Throwables can only be added as the last arg otherwise compilation
  will fail, this no longer depends on error-prone to validate
  arguments.

## After this PR
```diff
- Logger log = LoggerFactory.getLogger(Foo.class);
+ SafeLogger log = SafeLoggerFactory.getLogger(Foo.class);
log.info("Hello, {}!", SafeArg.of("whom", "World"));
```
==COMMIT_MSG==
Implement the SafeLogger API
==COMMIT_MSG==

## Possible downsides?
More logging APIs!

### CODE REVIEW

File: logger/src/main/java/com/palantir/logsafe/logger/SafeLoggerFactory.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+package com.palantir.logsafe.logger;
+
+import com.palantir.logsafe.logger.spi.SafeLoggerFactoryBridge;
+import java.util.Comparator;
+import java.util.ServiceLoader;
+
+/** Factory used to access {@link SafeLogger} instance. */
+public final class SafeLoggerFactory {
```
Comment (carterkozak) [inline logger/src/main/java/com/palantir/logsafe/logger/SafeLoggerFactory.java:29]: Public API

File: logger/src/main/java/com/palantir/logsafe/logger/SafeLogger.java [hunk 2]
```diff
@@ -0,0 +1,1647 @@
...
+import java.util.List;
+import java.util.Objects;
+import javax.annotation.processing.Generated;
+
+/**
+ * Safe-logging logger API used to produce log events with safe-classified {@link Arg} values.
+ */
+@Generated("com.palantir.logsafe.logger.generator.LoggerGenerator")
+@SuppressWarnings("TooManyArguments")
+public final class SafeLogger {
```
Comment (carterkozak) [inline logger/src/main/java/com/palantir/logsafe/logger/SafeLogger.java:28]: Public API

File: logger-generator/src/main/java/com/palantir/logsafe/logger/generator/LoggerGenerator.java [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+                    .map(param -> CodeBlock.of("$N", param.name))
+                    .collect(CodeBlock.joining(", "));
+            methodBuilder
+                    .addModifiers(Modifier.PUBLIC)
+                    .addStatement("$L$N.$N($L)", isVoidMethod ? "" : "return ", DELEGATE, spec.name, args);
+            loggerBuilder.addMethod(methodBuilder.build());
+        }
+        return JavaFile.builder(LOGGER_NAME.packageName(), loggerBuilder.build())
+                .skipJavaLangImports(true)
+                .addFileComment(copyright(2021))
```
Comment (ferozco) [inline logger-generator/src/main/java/com/palantir/logsafe/logger/generator/LoggerGenerator.java:106]: pedantic nit: use a block comment instead :P

**Review by ferozco — COMMENTED** [review]
I like the idea! excited for people to have to think less about their logs

File: logger-generator/src/main/java/com/palantir/logsafe/logger/generator/LoggerGenerator.java (hunk 3 above)
Comment (carterkozak) [reply logger-generator/src/main/java/com/palantir/logsafe/logger/generator/LoggerGenerator.java:106]: Any idea how? I tried to find an API for javapoet block file comments, but didn't have any luck. Closest thing was class-level jadadoc, but that isn't quite right either.

File: logger/src/main/java/com/palantir/logsafe/logger/SafeLoggerFactory.java [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+
+    private static final SafeLoggerFactoryBridge BRIDGE =
+            ServiceLoader.load(SafeLoggerFactoryBridge.class, SafeLoggerFactory.class.getClassLoader()).stream()
+                    .map(ServiceLoader.Provider::get)
+                    .max(Comparator.comparing(SafeLoggerFactoryBridge::priority))
+                    .orElse(DefaultSlf4JSafeLoggerFactoryBridge.INSTANCE);
+
+    /** Returns a {@link SafeLogger} for the {@code clazz} origin. */
+    public static SafeLogger get(Class<?> clazz) {
```
Comment (carterkozak) [inline logger/src/main/java/com/palantir/logsafe/logger/SafeLoggerFactory.java]:
```suggestion
    public static SafeLogger get(@Safe Class<?> clazz) {
```

File: logger/src/main/java/com/palantir/logsafe/logger/SafeLoggerFactory.java [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+        return new SafeLogger(BRIDGE.get(clazz));
+    }
+
+    /** Returns a {@link SafeLogger} for the {@code name} origin. */
+    public static SafeLogger get(String name) {
```
Comment (carterkozak) [inline logger/src/main/java/com/palantir/logsafe/logger/SafeLoggerFactory.java]:
```suggestion
    public static SafeLogger get(@Safe String name) {
```

**Comment by carterkozak** [comment]
I refactored out the slf4j binding into another module, which is a runtime dep of the logger module. Functionally equivalent, but it decouples things in a nicer way imo, and gives us more flexibility in the future.

File: logger/src/main/java/com/palantir/logsafe/logger/SafeLoggerFactory.java [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+                    // Should never happen unless dependencies are explicitly excluded
+                    .orElseThrow(NoSafeLoggerImplementationsException::new);
+
+    /** Returns a {@link SafeLogger} for the {@code clazz} origin. */
+    public static SafeLogger get(@Safe Class<?> clazz) {
+        return new SafeLogger(BRIDGE.get(clazz));
+    }
+
+    /** Returns a {@link SafeLogger} for the {@code name} origin. */
+    public static SafeLogger get(@Safe String name) {
```
Comment (carterkozak) [inline logger/src/main/java/com/palantir/logsafe/logger/SafeLoggerFactory.java:44]: We could enforce compile-time-constant strings here, or remove the string option entirely. I think `@Safe` is sufficient,and haven't encountered unsafe leaks through the logger name myself, so I don't want to make it difficult to implement things like https://github.com/palantir/tritium/blob/b3995307803a97cd2b5e1c5abc1d463070cac181/tritium-annotations/src/main/java/com/palantir/tritium/annotations/internal/InstrumentationBuilder.java#L78-L80
```suggestion
    public static SafeLogger get(@Safe @CompileTimeConstant String name) {
```

Comment (tpetracca) [reply logger/src/main/java/com/palantir/logsafe/logger/SafeLoggerFactory.java:44]: Agreed that compile time constant here is likely desirable, but not required, and would be a break in a few places. The risk here is that we propagate an Unsafe origin field (which is assumed safe) but we already hold that risk and it has never occurred (and highly unlikely to).

**Review by tpetracca — APPROVED** [review]
reviewed and discussed at length over vtc

**Pull request merged**